libtenzir/src/diagnostics.cpp

```diff
@@ -242,4 +243,26 @@ void diagnostic_builder::emit(const shared_diagnostic_handler& diag) && {
   return diag.emit(std::move(result_));
 }
 
+auto diagnostic_deduplicator::insert(const diagnostic& d) -> bool {
+  // We remember whether we have seen a diagnostic by storing its main message
+  // and the locations of its annotations.
+  // TODO: Improve this.
+  auto locations = std::vector<location>{};
+  for (auto& annotation : d.annotations) {
+    locations.push_back(annotation.source);
```

Comment:
Should the locations be sorted here?

Reply:
This would only matter if there is a diagnostic with multiple annotations where the order is derived from something else or even is non-deterministic. And then it's not clear to me whether you'd want to consider those duplicates.

Reply:
Makes sense.